src/runtime/runtime.ts

```diff
@@ -69,6 +81,106 @@ export class Interner {
 
 export var GlobalInterner = new Interner();
 
+//------------------------------------------------------------------------
+// EAVNs
+//------------------------------------------------------------------------
+
+type EAVNField = "e"|"a"|"v"|"n";
+class EAVN {
+  constructor(public e:ID, public a:ID, public v:ID, public n:ID) {}
+};
+
+//------------------------------------------------------------------------
+// Values
+//------------------------------------------------------------------------
+
+type ResolvedValue = ID|undefined|IgnoreRegister;
+
+//------------------------------------------------------------------------
+// Proposal
+//------------------------------------------------------------------------
+
+interface Proposal {
+  cardinality:number,
+  forFields:EAVNField[],
+  forRegisters:Register[],
+  proposer:Constraint,
+  skip?:boolean,
+  info?:any,
+}
+
+//------------------------------------------------------------------------
+// Indexes
+//------------------------------------------------------------------------
+
+interface Index {
+  insert(change:Change):void;
+  propose(proposal:Proposal, e:ResolvedValue, a:ResolvedValue, v:ResolvedValue, n:ResolvedValue, transaction:number, round:number):Proposal;
+  resolveProposal(proposal:Proposal):any[][];
+  get(e:ResolvedValue, a:ResolvedValue, v:ResolvedValue, n:ResolvedValue, transaction:number, round:number):EAVN[];
+}
+
+class ListIndex implements Index {
+  changes: Change[] = createArray();
+  insert(change:Change) {
+     this.changes.push(change);
+  }
+
+  resolveProposal(proposal:Proposal) {
+    return proposal.info;
+  }
+
+  propose(proposal:Proposal, e:ResolvedValue, a:ResolvedValue, v:ResolvedValue, n:ResolvedValue, transaction = Infinity, round = Infinity) {
+    let final = createArray() as ID[][];
+    let forFields:EAVNField[] = createArray();
+    let seen = createHash();
+
+    if(a === undefined) forFields.push("a");
+    else if(v === undefined) forFields.push("v");
+    else if(e === undefined) forFields.push("e");
+    else if(n === undefined) forFields.push("n");
+
+    for(let change of this.changes) {
+      if((e === undefined || e === IGNORE_REG || e === change.e) &&
+         (a === undefined || a === IGNORE_REG || a === change.a) &&
+         (v === undefined || v === IGNORE_REG || v === change.v) &&
+         (n === undefined || n === IGNORE_REG || n === change.n) &&
+         (change.transaction <= transaction) &&
+         (change.round <= round)) {
+        let current = change[forFields[0]];
+        if(!seen[current]) {
+          seen[current] = true;
+          final.push([current]);
+        }
+      }
+    }
+
+    proposal.cardinality = final.length;
+    proposal.info = final;
+    proposal.forFields = forFields;
```

Comment:
If this proposal solves only for `forFields[0]`, why are all variable fields returned in the proposal?